src/repository/opamDownload.ml

```diff
@@ -133,7 +119,58 @@ let download_command ~compress ?checksum ~url ~dst () =
   in
   let stdout = OpamSystem.temp_file ~auto_clean:false "dl" in
   OpamProcess.Job.finally (fun () -> OpamSystem.remove_file stdout) @@ fun () ->
-  OpamSystem.make_command ~allow_stdin:false ~stdout cmd args @@> tool_return url
+  OpamSystem.make_command ~allow_stdin:false ~stdout cmd args @@> c
+
+let tool_return redownload_command url ret =
+  match Lazy.force OpamRepositoryConfig.(!r.download_tool) with
+  | _, `Default ->
+    if OpamProcess.is_failure ret then
+      fail (Some "Download command failed",
+                Printf.sprintf "Download command failed: %s"
+                  (OpamProcess.result_summary ret))
+    else Done ()
+  | _, `Curl ->
+    if OpamProcess.is_failure ret then
+      if ret.r_code = 43 then begin
+        (* Code 43 is CURLE_BAD_FUNCTION_ARGUMENT (7.1 7-Aug-2000). This should
+           never be encountered using the curl binary, so we assume that it's
+           a manifestation of curl/curl#13845 (see also #6120). *)
+        log "Attempting to mitigate curl/curl#13845";
+        redownload_command ~with_curl_mitigation:true @@ function ret ->
+          if OpamProcess.is_failure ret then
+            if ret.r_code = 22 then
+              (* If this broken version of curl persists for some time, it is
+                 relatively straightforward to parse the http response code from
+                 the message, as it hasn't changed. *)
+              fail (Some "cURL failed owing to a server-side issue",
+                      Printf.sprintf "cURL failed with server-side error: %s"
+                        (OpamProcess.result_summary ret))
+            else
+              fail (Some "cURL failed",
+                      Printf.sprintf "cURL failed: %s"
+                        (OpamProcess.result_summary ret))
+          else Done ()
+      end else
+        fail (Some "Curl failed", Printf.sprintf "Curl failed: %s"
+                    (OpamProcess.result_summary ret))
+    else
+      match ret.OpamProcess.r_stdout with
+      | [] ->
+        fail (Some "curl empty response",
+                  Printf.sprintf "curl: empty response while downloading %s"
+                    (OpamUrl.to_string url))
+      | l  ->
+        let code = List.hd (List.rev l) in
+        let num = try int_of_string code with Failure _ -> 999 in
+        if num >= 400 then
+          fail (Some ("curl error code " ^ code),
+                    Printf.sprintf "curl: code %s while downloading %s"
```

Comment:
```suggestion
                    Printf.sprintf "cURL: code %s while downloading %s"
```

Reply:
The proper name is curl: https://curl.se/

Reply:
oh ok